exporter/googlecloudpubsubexporter/config.go

```diff
@@ -34,18 +35,66 @@ type Config struct {
 	ProjectID string `mapstructure:"project"`
 	// User agent that will be used by the Pubsub client to connect to the service
 	UserAgent string `mapstructure:"user_agent"`
-	// Override of the Pubsub endpoint
-	Endpoint string `mapstructure:"endpoint"`
+	// Override of the Pubsub endpoint, for testing only
+	endpoint string
 	// Only has effect if Endpoint is not ""
-	Insecure bool `mapstructure:"insecure"`
+	insecure bool
 
 	// The fully qualified resource name of the Pubsub topic
 	Topic string `mapstructure:"topic"`
+	// Compression of the payload (only gzip or is supported, no compression is the default)
+	Compression string `mapstructure:"compression"`
+	// Watermark defines the watermark (the ce-time attribute on the message) behavior
+	Watermark WatermarkConfig `mapstructure:"watermark"`
+}
+
+// WatermarkConfig customizes the behavior of the watermark
+type WatermarkConfig struct {
+	// Behavior of the watermark. Currently, only  of the message (none, earliest and current, current being the default)
+	// will set the timestamp on pubsub based on timestamps of the events inside the message
+	Behavior string `mapstructure:"behavior"`
+	// Indication on how much the timestamp can drift from the current time, the timestamp will be capped to the allowed
+	// maximum. A duration of 0 is the same as maximum duration
+	AllowedDrift time.Duration `mapstructure:"allowed_drift"`
 }
 
 func (config *Config) Validate() error {
 	if !topicMatcher.MatchString(config.Topic) {
-		return fmt.Errorf("topic '%s' is not a valide  format, use 'projects/<project_id>/topics/<name>'", config.Topic)
+		return fmt.Errorf("topic '%s' is not a valid format, use 'projects/<project_id>/topics/<name>'", config.Topic)
+	}
+	_, err := config.parseCompression()
+	if err != nil {
+		return err
+	}
+	return config.Watermark.validate()
+}
+
+func (config *WatermarkConfig) validate() error {
+	if config.AllowedDrift == 0 {
+		config.AllowedDrift = 1<<63 - 1
+	}
+	_, err := config.parseWatermarkBehavior()
+	return err
+}
+
+func (config *Config) parseCompression() (compression, error) {
+	switch config.Compression {
+	case "gzip":
+		return gZip, nil
+	case "":
```

Comment:
nit: it might be nice to have an explicit "uncompressed" value, and default to that in createDefaultConfig().  That way you can separate defaulting vs parsing of config.